cpp/src/reductions/simple.cuh

```diff
@@ -0,0 +1,162 @@
+            *dcol, Op::Op::template identity<ElementType>()),
+        typename Op::template transformer<ResultType>{});
+    detail::reduce(dev_result.data(), it, col.size(), identity,
+                   typename Op::Op{}, mr, stream);
+  } else {
+    auto it = thrust::make_transform_iterator(
+        dcol->begin<ElementType>(),
+        typename Op::template transformer<ResultType>{});
+    detail::reduce(dev_result.data(), it, col.size(), identity,
+                   typename Op::Op{}, mr, stream);
+  }
+
+  using ScalarType = cudf::experimental::scalar_type_t<ResultType>;
+  auto s = new ScalarType(dev_result.value(), true, stream, mr);
+  std::unique_ptr<scalar> result = std::unique_ptr<scalar>(s);
+
+  // set scalar is valid
+  if (col.null_count() < col.size())
+    result->set_valid(true, stream);
+  else
+    result->set_valid(false, stream);
+  return result;
+};
+
+// @brief result type dispatcher for simple reduction (a.k.a. sum, prod, min...)
+template <typename ElementType, typename Op>
+struct result_type_dispatcher {
+private:
+    template <typename ResultType>
+    static constexpr bool is_supported_v()
+    {
+      // for single step reductions,
+      // the available combination of input and output dtypes are
+      //  - same dtypes (including cudf::wrappers)
+      //  - any arithmetic dtype to any arithmetic dtype
+      //  - cudf::experimental::bool8 to/from any arithmetic dtype
+      return std::is_convertible<ElementType, ResultType>::value &&
+             (std::is_arithmetic<ResultType>::value ||
+              std::is_same<Op, cudf::experimental::reduction::op::min>::value ||
+              std::is_same<Op, cudf::experimental::reduction::op::max>::value);
+    }
+
+public:
+    template <typename ResultType, std::enable_if_t<is_supported_v<ResultType>()>* = nullptr>
+    std::unique_ptr<scalar> operator()(column_view const& col, data_type const output_dtype,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+    {
+      return simple_reduction<ElementType, ResultType, Op>(col, output_dtype, mr, stream);
+    }
+
+    template <typename ResultType, std::enable_if_t<not is_supported_v<ResultType>()>* = nullptr>
+    std::unique_ptr<scalar> operator()(column_view const& col, data_type const output_dtype,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+    {
+        CUDF_FAIL("input data type is not convertible to output data type");
+    }
+};
+
+// @brief input column element for simple reduction (a.k.a. sum, prod, min...)
+template <typename Op>
+struct element_type_dispatcher {
+private:
+    // return true if ElementType is arithmetic type or bool8, or
+    // Op is DeviceMin or DeviceMax for wrapper (non-arithmetic) types
+    template <typename ElementType>
+    static constexpr bool is_supported_v()
+    {
+      // disable only for string ElementType except for operators min, max, sum.
+      return  !( std::is_same<ElementType, cudf::string_view>::value &&
```

Comment:
If this check is inside function calls, compilation will try to compile for all types in type_dispatcher. It won't compile because some types don't have `operator+`, etc.
AFAIK, this needs to be SFINAE. I don't see a way to move to individual reduction function calls. any suggestions?